### PR TITLE
New handcrafted features

### PR DESCRIPTION
**Pull request by pknut**
* average monthly change on a credit card (credit_card_monthly_diff_mean)
* number of remaining installments (pos_cash_remaining_installments)
* number of completed contracts (pos_cash_completed_contracts)

### CODE REVIEW

File: src/feature_extraction.py [hunk 1]
```diff
@@ -340,6 +340,62 @@ def fit(self, X, credit_card, **kwargs):
         features['credit_card_cash_card_ratio'] = features['credit_card_drawings_atm'] / features[
             'credit_card_drawings_total']
 
+        credit_card_sorted = credit_card.sort_values(['SK_ID_CURR', 'MONTHS_BALANCE'])
```
Comment (jakubczakon) [inline src/feature_extraction.py]: @pknut so this is the mean month to month difference of credit card balance? 
We could probably add say last 5 month credit balance as 5 features or maybe those diffs as 5 (or 10 idk) features

File: src/feature_extraction.py [hunk 2]
```diff
@@ -340,6 +340,62 @@ def fit(self, X, credit_card, **kwargs):
...
+    @property
+    def feature_names(self):
+        feature_names = list(self.features.columns)
+        feature_names.remove('SK_ID_CURR')
+        return feature_names
+
+    def fit(self, X, pos_cash, **kwargs):
+        features = pd.DataFrame({'SK_ID_CURR': pos_cash['SK_ID_CURR'].unique()})
+
+        pos_cash_sorted = pos_cash.sort_values(['SK_ID_CURR', 'MONTHS_BALANCE'])
```
Comment (jakubczakon) [inline src/feature_extraction.py]: @pknut maybe something like remaining_installment / remaining_months or / monthly_payment could be another feature here

File: src/feature_extraction.py [hunk 3]
```diff
@@ -340,6 +340,62 @@ def fit(self, X, credit_card, **kwargs):
...
+        group_object = pos_cash_sorted.groupby('SK_ID_CURR')['CNT_INSTALMENT_FUTURE'].last().reset_index()
+        group_object.rename(index=str,
+                            columns={'CNT_INSTALMENT_FUTURE': 'pos_cash_remaining_installments'},
+                            inplace=True)
+        features = features.merge(group_object, on=['SK_ID_CURR'], how='left')
+
+        pos_cash['is_contract_status_completed'] = pos_cash['NAME_CONTRACT_STATUS'] == 'Completed'
```
Comment (jakubczakon) [inline src/feature_extraction.py]: @pknut If we have number of completed contracts than why not add fraction of completed contracts?

**Comment by jakubczakon** [comment]
@pknut Also I think it is a good idea to encapsulate those dynamic features in a _dynamic_features(self, X) and have _static_features(self, X) for the rest. by doing so we would know what is what and we could easily turn on/off stuff.

File: src/feature_extraction.py (hunk 1 above)
Comment (pknut) [reply src/feature_extraction.py]: Unfortunately after adding these new ideas (last 5 months credit balance as 5 features and an average of 5 last months diffs) our local score decrease.

Comment (jakubczakon) [reply src/feature_extraction.py]: interesting, more investigation needed i guess

**Pull request merged**